### DISCUSSION TITLE
save on exit not working

### DISCUSSION THREAD

**Discussion by lord-muad-dib** [ramokz/phantom-camera#403] [Question / Help Wanted] [answered]
hello, i understand it might sound silly but, whenever i enable this plugin (even on a blank project, no need to even add a phantom camera to the scene) the save on exit function of the editor stops working.
like.. i edit something, then i press the X to close the editor, the popup shows up, but when i press save.. godot closes without saving.
tested it on windows 10 and 11 on godot 4.3 and 4.4. edit: plugin version 7.3
one thing i noticed is that everytime i start the editor i get a wildcard on top, like something's modified (which might be something with the autoload adding itself each time.. i suppose?) it doesn't happen with other plugins

note that manual saving and the autosave before running the project are still working
and
if i disable the plugin and restart. the save on exit starts working again, and this behaviour doesn't happen with other plugins

i feel stupid, because i feel this is something basic that everyone should see immidiatly it's not working, so maybe it's something that i do wrong? that's why i didn't file a bug issue
but then again, i tried with a blank editor config and project, same problem

thanks

**Comment by ramokz** [comment]
It sounds very odd.
I am honestly not sure what it could be, as I have never come across the issue before or heard anyone experiencing something similar.

My only guess would be that Godot doesn't have full write permission where it's installed.
How did you install Godot?

**Comment by lord-muad-dib** [comment]
thanks for posting!
i know right?
yeah well, just unzipping godot in a folder i have full access to should do the trick. same for the projects.
manual saving (ctrl+s) and the automatic saving before running a scene keeps working anyways.

also tried checking with godot-console, or running it from a terminal to see if there are any relevant messages left behind and... nope, very strange

is the wildcard on editor startup normal?

i'm on godot cafe, if you want we can DM and check things together

**Reply by ramokz** [reply]
Any random behavior when starting up the editor isn't normal, no. It should be persistent every time it opens.

I sadly don't really have time to help out individuals if an issue appears to be exclusive to their machine — as it often tends to be either hardware or software/OS-specific complications.

Got quite a backlog of bugs to resolve that affects multiple people, which I need to prioritize, unfortunately...

**Reply by lord-muad-dib** [reply]
this happens on two different machines but err.. i'm just trying to be helpfull..

ok i'll try to investigate by myself first and see if i catch on something, when i have time i'll try to downgrade stuff and bisect commits

**Reply by ramokz** [reply]
Sorry, didn't mean to imply that I don't want to help or that your issue isn't important or concerning.

What I meant is that I have a couple of dozens of issues to resolve that I can verify and replicate myself, so the solutions for those are a lot more tangible. I only have a few hours every other day outside of work, and when energy allows, to work on this project. So I want to maximize the time I spend fixing / improving things that affect as many people as possible and with the least amount of time spent figuring out what causes the issue. If I did this full-time, it would be a different story.

A general rule of thumb of issue reporting is that if an issue cannot be replicated by others or if there are no error logs, then it's tricky to justify prioritizing that issue over others that can be consistently reproduced. There have been a few individual specific issues with strange behaviors popping up over time (recently #405), which you can probably imagine can quickly take up most of the time I have available on the project.

If you can narrow down where the issue might lie, then I am happy to try to help out.

**Reply by lord-muad-dib** [reply]
that's ok i understand :) developing on spare time is already hard as is
thankfully it's not a blocking issue per se, i just have to remember to save manually each time eheh

and sure, i will investigate the issue. it seems to be a challenging one. i'll take the chance to learn more about the editor's internals, i hope to find some way to trace what godot is doing on startup
maybe some behaviours changed between versions of godot

for now ill keep this thread open for other users to intervene if they want, i'll probably open an issue once i start digging

thanks again

**Reply by ramokz** [reply]
Thanks for understanding, and yes, do share any findings you come across and we can try and figure something out :)

**Reply by lord-muad-dib** [reply]
ok, i figured it out, sorry if i didn't get you back sooner, didn't have time to check it out

looking in plugin.gd
```
func _enter_tree() -> void

	add_autoload_singleton(PHANTOM_CAMERA_MANAGER, "res://addons/phantom_camera/scripts/managers/phantom_camera_manager.gd") #<--this line
```

first of all, a bit of housekeeping :). this line is called 2 times. the other line is at the end of the method. it doesn't seem to hurt but it prints added autoload two times in the console. doesn't seem there's the need to have 2 identical calls, i deleted one and it's still working.

now for the real fix.
it seems that that line above doesn't check if the autoload is already in the list of available autoloads and tries to re-add it again, that's why the editor is pushing edits to the project resources. it shows the wildcard at the top left of the window, and the save starts misbehaving untill you save the project manually once.... but the issue comes back once you reload it.

i've fixed it like this
honestly i don't know if there's a better way to check user-added singletons/autoload yet, but it works, 
```
if not ProjectSettings.has_setting("autoload/"+str(PHANTOM_CAMERA_MANAGER)): #check if the autoload isn't already added in the settings

	add_autoload_singleton(PHANTOM_CAMERA_MANAGER, "res://addons/phantom_camera/scripts/managers/phantom_camera_manager.gd")
```
i hope it helps
thank you

edit: after a bit of research, i'm coming to the conclusion that add_autoload_singleton is probably buggy...
i've been testing other plugins that add autoloads and they all have the same issue and my workaround with the check works on all of them.
i don't think it's possible that no one noticed this thing :)

**Reply by ramokz** [reply]
Thanks for reporting back with your findings!

Yeah, autoload is generally a bit messy when working with addons/plugins. E.g., unlike the end user, you cannot reference the autoloaded class inside the addon, as that will cause a reference error — hence why it is being referenced using `_phantom_camera_manager = get_tree().root.get_node(_constants.PCAM_MANAGER_NODE_NAME)` in various places.

You _should_ be able to achieve the same result with a similar pattern as the rest of the addon:
```gdscript
if not get_tree().root.get_node_or_null(String(PHANTOM_CAMERA_MANAGER)):
  add_autoload_singleton(PHANTOM_CAMERA_MANAGER, "res://addons/phantom_camera/scripts/managers/phantom_camera_manager.gd")
```

Fundamentally, I didn't see the need for adding in a checker, but if that solves your issue, then I'm happy to include that as it does no harm either. I suspect many have experienced the error, but no one has bothered to report it.

If you want to be credited on the repo for the finding, and solution, then do also feel free to submit a PR :) Will be including it in the next release either way :)

**Reply by lord-muad-dib** [reply]
thanks! yeah, checking directly what's currently loaded on the tree is probably a better solution.
i've made more tests, and the save misbehaving with the wildcard happens in easier ways to debug and if you can test it we can report the issue to godot.
try to manually uncheck the autoload. you will see the wildcard on the topleft, don't save or run the project yet. try to close the editor. godot should complain about modified resources.
now, abort and go back to the editor.
at this point edit a gdscript and then try saving on exit like before. godot will not recognize there's an edited gdscript. if you save at the prompt and then restart godot.. the edits on the gdscript should be lost.

if you can test this with me it would be awesome!

at this time i'm unable to do a PR. i'm away home for few days, but i hope you can release 8.1 asap, a small shoutout is fine! i'm sure i'll be able to contribute in more ways soon enough :) this plugin is great

**Reply by ramokz** [reply, marked as answer]
So I couldn't replicate the exact issue you described, but the editor does, to put it mildly, freak out and effectively freeze if you uncheck the autoloaded file within `Project Settings` for me. It's due to the various scripts that reference the singleton panic as they lose access to the class, which is to be expected really.

Autoloaded scripts and plugins do not gel well together in general, so am keen on seeing better synergy some time in the future. Though, that's a larger and more complicated conversation, I reckon.

I can add in the change we talked about, with a credit note, for the next release if you're fine with that. Suspect the next release will be out in a few days, so if you change your mind about not getting PR credit do let me know! :)

**Comment by lord-muad-dib** [comment]
yeah i'm totally ok of course, no need to wait, thanks ;)

i've been reaching out to other people and some do have this issue, some don't and some have different behaviours altogether lmao.

it's very bizarre it seems to be both OS dependant and random at the same time. i'll try to file an issue to godot once i'm able to, let's see how it goes, heh.

thanks again

**Reply by ramokz** [reply]
Yeah, have encountered a few of those issues over time; always a nightmare to debug.

Thanks again for reporting the issue and for following up on it!

**Comment by lord-muad-dib** [comment]
let's see if this  will be merged for 4.5 godotengine/godot/pull/109863 a proper fix to add_autoload_singleton checking if stuff is already present. whenever you'll change the minimum requirement of PC you could delete the workaround ;)

**Reply by ramokz** [reply]
Thanks for following up on this and sharing the issue link!

Looked up the workaround in the codebase we discussed above, and it looks like it's actually no longer being used in the current version. So there doesn't seem to be any cleanup to be done if that PR gets merged, just a slightly better DX for end users ;)

**Reply by lord-muad-dib** [reply]
oh, i didn't even notice (though, i guess you could still delete the condition on line 47 in the future) that this bug could be back to the plugin because i have tons of plugins that still triggered the same issue anyways.. haha.

**Reply by ramokz** [reply]
The checker on line 47 is mainly to prevent adding the singleton multiple times if a user disables and then re-enables the addon. 
Not 100% sure it's mission critical to have, but a single extra line isn't a noticeable complication to add, so don't really have an issue with it.

Checkers like these are sometimes needed and sometimes not, it seems a bit random tbh, so it's mostly a consistency thing to guard against the odd bug that may occur from not having it. Might explore stripping things like these out and re-add them only as needed in the future.
